### PR TITLE
Update sparsity check heuristics used in TorchGather

### PR DESCRIPTION
**Pull request by yeounoh**
Addresses #3587

### CODE REVIEW

**Comment by JackCaoG** [comment]
@wonjoolee95 FYI, we will need this pr in the 1.12. You can still play with trigger for now but we will likely need to rebuild the rc(maybe rc1?) with this fix.

@yeounoh Can you make sure this change does not regress vit model?

**Comment by yeounoh** [comment]
> @wonjoolee95 FYI, we will need this pr in the 1.12. You can still play with trigger for now but we will likely need to rebuild the rc(maybe rc1?) with this fix.
> 
> @yeounoh Can you make sure this change does not regress vit model?

Ok, just verified the new heuristics with the ViT model from @ronghanghu . 

@ymwangg , this would allow sparse gather on GPU and CPU -- thank you again for your detailed analysis.

File: torch_xla/csrc/ops/gather.cpp [hunk 1]
```diff
@@ -7,31 +7,44 @@
 #include "torch_xla/csrc/ops/infer_output_shape.h"
 
 namespace torch_xla {
+namespace {
+
+xla::Shape NodeOutputShape(const torch::lazy::Value& input,
```
Comment (JackCaoG) [inline torch_xla/csrc/ops/gather.cpp:12]: hmm, will `IsSparseGather` change the output shape? Why do we need a `NodeOutputShape` here?

**Review by JackCaoG — COMMENTED** [review]
Mostly lGTM, one minor question.

File: torch_xla/csrc/ops/gather.cpp (hunk 1 above)
Comment (yeounoh) [reply torch_xla/csrc/ops/gather.cpp:12]: It does, since dense gather actually uses reduce instead of gather.

**Review by JackCaoG — APPROVED** [review]
Thanks!

File: torch_xla/csrc/ops/gather.cpp (hunk 1 above)
Comment (ymwangg) [reply torch_xla/csrc/ops/gather.cpp:12]: FYI https://pytorch.org/docs/stable/generated/torch.gather.html. The output tensor of `torch.gather` has the same shape as `index`.

An issue I found with XLA dense gather is that it does not allow `index.size(d) < input.size(d)` for dimension `d != dim` so that the following code does not work with dense gather. Perhaps this type of input is not typical since nobody has complained.
```
a = torch.randint(0, 10, size=(8, 8), dtype=torch.int64, device=device)
b = torch.ones(10, 10).to(device)
c = torch.gather(b, dim=1, index=a)
```

Comment (yeounoh) [reply torch_xla/csrc/ops/gather.cpp:12]: I see, @ymwangg let me check with XLA folks about this before landing, and see if they could modify the implementation or the check(s).

Comment (yeounoh) [reply torch_xla/csrc/ops/gather.cpp:12]: @ymwangg I think it's a bug and will file a separate bug to track internally. Thanks!

**Comment by yeounoh** [comment]
Merging cc @JackCaoG @ymwangg

**Comment by JackCaoG** [comment]
@yeounoh Thanks, Can you backport this pr to r1.12?
